### DISCUSSION TITLE
Problems with NetCDF library when compiling MET

### DISCUSSION THREAD

**Discussion by stephentjemkes** [dtcenter/METplus#2086] [Installation] [answered]
*Replace italics below with details for this issue.*

## Describe the Problem ##
Can not compile as it can not find the #include <netcdf> file

Making all in vx_nc_util
make[4]: Entering directory '/home/stephen/BlackForestWeather/MET-11.0.1/src/libcode/vx_nc_util'
g++ -DHAVE_CONFIG_H -I. -I../../..  -I../../../src/basic/vx_cal -I../../../src/basic/vx_config -I../../../src/basic/vx_log -I../../../src/basic/vx_math -I../../../src/basic/vx_util -I../../../src/basic/vx_util_math -I../../../src/libcode/vx_afm -I../../../src/libcode/vx_analysis_util -I../../../src/libcode/vx_color -I../../../src/libcode/vx_data2d -I../../../src/libcode/vx_data2d_factory -I../../../src/libcode/vx_data2d_grib -I../../../src/libcode/vx_data2d_grib2 -I../../../src/libcode/vx_data2d_nccf -I../../../src/libcode/vx_data2d_nc_met -I../../../src/libcode/vx_data2d_nc_pinterp -I../../../src/libcode/vx_data2d_python -I../../../src/libcode/vx_python3_utils -I../../../src/libcode/vx_geodesy -I../../../src/libcode/vx_gis -I../../../src/libcode/vx_gnomon -I../../../src/libcode/vx_grid -I../../../src/libcode/vx_gsl_prob -I../../../src/libcode/vx_nav -I../../../src/libcode/vx_nc_obs -I../../../src/libcode/vx_nc_util -I../../../src/libcode/vx_pb_util -I../../../src/libcode/vx_plot_util -I../../../src/libcode/vx_pointdata_python -I../../../src/libcode/vx_ps -I../../../src/libcode/vx_pxm -I../../../src/libcode/vx_render -I../../../src/libcode/vx_regrid -I../../../src/libcode/vx_shapedata -I../../../src/libcode/vx_solar -I../../../src/libcode/vx_statistics -I../../../src/libcode/vx_stat_out -I../../../src/libcode/vx_bool_calc -I../../../src/libcode/vx_summary -I../../../src/libcode/vx_time_series -I../../../src/libcode/vx_series_data -I../../../src/libcode/vx_seeps -I../../../src/libcode/vx_tc_util -fopenmp -I/home/stephen/local/include -I/home/stephen/local/include -I/home/stephen/local/include -DBLOCK4 -DMET_BASE="\"/home/stephen/local/share/met\""   -MT libvx_nc_util_a-nc_var_info.o -MD -MP -MF .deps/libvx_nc_util_a-nc_var_info.Tpo -c -o libvx_nc_util_a-nc_var_info.o `test -f 'nc_var_info.cc' || echo './'`nc_var_info.cc
nc_var_info.cc:15:10: fatal error: netcdf: No such file or directory
   15 | #include <netcdf>

the listing of the /home/stephen/local/include shows not netcdf but the netcdf.h and netcdf.inc

so not clear why the code expects a netcdf file in the include subdirectory 

cd /home/stephen/local/include

ls *netcdf*
netcdf4_f03.mod            netcdf4_nf_interfaces.mod  netcdf_dispatch.h  netcdf_filter_build.h  netcdf_fortv2_c_interfaces.mod  netcdf.inc    netcdf_meta.h  netcdf_nc_data.mod        netcdf_nf_data.mod
netcdf4_nc_interfaces.mod  netcdf_aux.h               netcdf_f03.mod     netcdf_filter.h        netcdf.h                        netcdf_mem.h  netcdf.mod     netcdf_nc_interfaces.mod  netcdf_nf_interfaces.mod

You can see the -I/home/stephen/local/include

so the environmental variables are set

### Expected Behavior ###
*Provide a clear and concise description of what you expected to happen here.*

### Environment ###
Describe your runtime environment:
*1. linux workstathon
*2. ubuntu 22.02
*3. almost latest versions of netcdf, hdf5 etc. 

please clarify what netcdf file needs to be included

**Comment by jprestop** [comment]
Hi @stephentjemkes.  

I see that you are having trouble compiling MET due to the error:
```
nc_var_info.cc:15:10: fatal error: netcdf: No such file or directory
```

The support for NetCDF version 4 in MET, requires NetCDF-C, NetCDF-CXX, and HDF5. This file, "netcdf", is part of the NetCDF CXX library, which is required in addition to the NetCDF-C library.  MET expects both of these libraries to be installed in the same location, as there is only one MET_NETCDF variable to set in the configuration of MET.  

The MET team has put together a compilation script to make it easier for users to compile MET and its dependencies.  Please take a look at the "Sample Script For Compiling External Libraries And MET" section of the [MET Download](https://dtcenter.org/community-code/model-evaluation-tools-met/download) webpage for how to get and use this script if you think it would be helpful.

Please let us know what other questions you have.

**Reply by JohnHalleyGotway** [reply]
@stephentjemkes, just following up on this discussion. Do any more questions remain or can we mark it as being answered?

**Reply by jprestop** [reply]
@stephentjemkes I just wanted to follow up on this Discussion again. Do you more questions remain or can we mark it as being answered?

**Comment by stephentjemkes** [comment]
Hi

This has not yet been resolved by me.

And i had to turn my attention to other topics.
If of interest to you can reactivate my effort next week in an attempt  to resolve this.
It is important to me as it might help to compare wrf with other datasets

Stephen

Sent from Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Reply by jprestop** [reply]
@stephentjemkes Thanks!  When you're able to try the compilation again, please let us know how it goes.

**Comment by stephentjemkes** [comment]
Hi,

I am re-compiling the met package (11.0.2), but there is now an issue with a dependency. In particular nceplibs-grib_util. despite cloning most recent versions of the nceplibs utilities, and latest jasper (4.0.0) the grib_util failed to compile.
I raised an issue on the github with the developing team to learn if this is an easy fix or not.
Note that ncep has pushed all its code to github so you may want to update the INSTALL text.


in the past Cairo was causing also problems and not sure if this is mandatory maybe they have updated their installation over the years. installing nceplibs has advanced (despite the awkward failure at the end 🙁)

I understand that the grib2 library is not mandatory,
the bufrlib was successfully installed so i will give it a go.

netcdf / hdf5 should be reasonable recent

stephen
________________________________

**Comment by stephentjemkes** [comment]
Hi,

I followed the instructions in the  INSTALL text

There are a few minor updates in this text needed.

=========================================

In the INSTALL text there is reference to

 http://dtcenter.org/community-code/model-evaluation-tools-met/online-tutorial

this does not exists anymore please replace by

https://dtcenter.org/community-code/model-evaluation-tools-met/documentation

and also the ncep libs are on github.

=====================


The issue is still the same the # include <netcdf> in the nc_var_info.cc code blocks compilation

I do not know enough of cc soi do not know if the code needs the netcdf.h or the netcdf.inc code or??

The -I/home/stephen/local/include is in the g++ statement and a listing (of the relevant part of that directory) shows that there are all sorts of netcdf files

netcdf.h netcdf.inc netcdf.mod

===========================================================================

-rw-r--r--  1 stephen stephen   1337 mei 20  2022 netcdf_mem.h
-rw-r--r--  1 stephen stephen  69042 mei 20  2022 netcdf.h
-rw-r--r--  1 stephen stephen   2292 mei 20  2022 netcdf_filter.h
-rw-r--r--  1 stephen stephen   9138 mei 20  2022 netcdf_filter_build.h
-rw-r--r--  1 stephen stephen  12573 mei 20  2022 netcdf_dispatch.h
-rw-r--r--  1 stephen stephen   2693 mei 20  2022 netcdf_aux.h
-rw-r--r--  1 stephen stephen    496 mei 20  2022 typesizes.mod
-rw-r--r--  1 stephen stephen  11482 mei 20  2022 netcdf_nf_interfaces.mod
-rw-r--r--  1 stephen stephen   9962 mei 20  2022 netcdf_nf_data.mod
-rw-r--r--  1 stephen stephen  21295 mei 20  2022 netcdf_nc_interfaces.mod
-rw-r--r--  1 stephen stephen   5810 mei 20  2022 netcdf_nc_data.mod
-rw-r--r--  1 stephen stephen  22588 mei 20  2022 netcdf.mod
-rw-r--r--  1 stephen stephen  24928 mei 20  2022 netcdf_fortv2_c_interfaces.mod
-rw-r--r--  1 stephen stephen  21932 mei 20  2022 netcdf_f03.mod
-rw-r--r--  1 stephen stephen   5559 mei 20  2022 netcdf4_nf_interfaces.mod
-rw-r--r--  1 stephen stephen  26834 mei 20  2022 netcdf4_nc_interfaces.mod
-rw-r--r--  1 stephen stephen  27137 mei 20  2022 netcdf4_f03.mod
-rw-r--r--  1 stephen stephen  61393 mei 20  2022 netcdf.inc

========================================================================

and the installation of the netcdf has been done using the standard approach.

nc-config --prefix => /home/stephen/local
nc-config --includedir => /home/stephen/local/include


my installation script now reads (enclosed)

========================================================================
#!/bin/ksh
logfile="make_install.log"
logconf="configure.log"


if [ -d $logconf ];
    then
      echo "$logconf exists";
      rm -rf $logconf;
fi
if [ -d $logfile ];
    then
      echo "$logfile exists";
      rm -rf $logfile;
fi


export MET_PYTHON_CC=`python3-config --cflags`
export MET_PYTHON_LD=`python3-config --ldflags`
export MET_HDF5=/home/stepen/local
export MET_GSL=/home/stephen/local
export MET_BUFR=/home/stephen/local
export MET_NETCDF=`nc-config --prefix`
./configure --prefix=/home/stephen/local --enable-python 2>&1 | tee $logconf
make install 2>&1 | tee $logfile
================================================================

but as indicated it fails (see make_install.log for the make results and the configure.log for configuration information)

================================================================

make[3]: Entering directory '/home/stephen/local/MET-11.0.2/src/libcode/vx_nc_util'
g++ -DHAVE_CONFIG_H -I. -I../../..  -I../../../src/basic/vx_cal -I../../../src/basic/vx_config -I../../../src/basic/vx_log -I../../../src/basic/vx_math -I../../../src/basic/vx_util -I../../../src/basic/vx_util_math -I../../../src/libcode/vx_afm -I../../../src/libcode/vx_analysis_util -I../../../src/libcode/vx_color -I../../../src/libcode/vx_data2d -I../../../src/libcode/vx_data2d_factory -I../../../src/libcode/vx_data2d_grib -I../../../src/libcode/vx_data2d_grib2 -I../../../src/libcode/vx_data2d_nccf -I../../../src/libcode/vx_data2d_nc_met -I../../../src/libcode/vx_data2d_nc_pinterp -I../../../src/libcode/vx_data2d_python -I../../../src/libcode/vx_python3_utils -I../../../src/libcode/vx_geodesy -I../../../src/libcode/vx_gis -I../../../src/libcode/vx_gnomon -I../../../src/libcode/vx_grid -I../../../src/libcode/vx_gsl_prob -I../../../src/libcode/vx_nav -I../../../src/libcode/vx_nc_obs -I../../../src/libcode/vx_nc_util -I../../../src/libcode/vx_pb_util -I../../../src/libcode/vx_plot_util -I../../../src/libcode/vx_pointdata_python -I../../../src/libcode/vx_ps -I../../../src/libcode/vx_pxm -I../../../src/libcode/vx_render -I../../../src/libcode/vx_regrid -I../../../src/libcode/vx_shapedata -I../../../src/libcode/vx_solar -I../../../src/libcode/vx_statistics -I../../../src/libcode/vx_stat_out -I../../../src/libcode/vx_bool_calc -I../../../src/libcode/vx_summary -I../../../src/libcode/vx_time_series -I../../../src/libcode/vx_series_data -I../../../src/libcode/vx_seeps -I../../../src/libcode/vx_tc_util -fopenmp -I/home/stephen/local/include -I/home/stepen/local/include -I../../basic/vx_log -I../../basic/vx_util -I/home/stephen/local/include/python3.10 -I/home/stephen/local/include/python3.10  -Wno-unused-result -Wsign-compare  -DNDEBUG -g -fwrapv -O3 -Wall -I/home/stephen/local/include -DBLOCK4 -DWITH_PYTHON -DMET_BASE="\"/home/stephen/local/share/met\""   -MT libvx_nc_util_a-nc_var_info.o -MD -MP -MF .deps/libvx_nc_util_a-nc_var_info.Tpo -c -o libvx_nc_util_a-nc_var_info.o `test -f 'nc_var_info.cc' || echo './'`nc_var_info.cc
nc_var_info.cc:15:10: fatal error: netcdf: No such file or directory
   15 | #include <netcdf>


===================================================================


Any suggestions would be appreciated.

thanks
stephen

________________________________

**Reply by jprestop** [reply, marked as answer]
Hi @stephentjemkes. The necessary and missing file, simply "netcdf" with no file extension, is part of the NetCDF CXX library, which is required in addition to the NetCDF-C library. My guess is that you have installed the NetCDF-C library but not the NetCDF-CXX library.  Both are required for MET.

I suggest taking a look at the "Sample Script For Compiling External Libraries And MET" section of the [MET Download](https://dtcenter.org/community-code/model-evaluation-tools-met/download) webpage for how to get and use this installation script that will install all of MET's dependencies and MET.

**Comment by stephentjemkes** [comment]
That could well be. It was not clear to me that i needed this netcdf library. Will install and try again thanks

Sent from Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by stephentjemkes** [comment]
Hi

As i have installed a lot of elements for my processor in my user environment to ensure that i have an stable environment, i have downloaded and installed a lot of stuff.
ndeed i have not yet installed the NetCDF-CXX library.

What i frequently encounter is that developers assume that users have sudo rights.
I would like to point out that "do not install in "/" unless you have to" would be a much better approach.
For instance i started to install the Cxx library but it failed using the basic setup as I had to indicate to  cmake to look for the hdf5.h header file in my local environment and not in the system environment.
But this was not enough as the met installation required c++ library and not cxx.
So in the end i used the script from your web page.

This script looks very interesting. I installed it all in a directory structure separate from my own user environment,  if i run it without any environmental variables pointing to my user environment it works nicely.

Maybe in the near future i will take your script and my own script and combine the two to have one environment, as there are lots of similarities, so it should not be too much time.

bottom line, i was able to compile and install using a basic version of the script.  Next is to start learning the tools to validate my assimilation experiments

many thanks for the support
stephen

________________________________

**Reply by jprestop** [reply]
Hi @stephentjemkes.  As a developer who does not have sudo access on my project's machines, I agree with your statements, which is why we allow the user's to provide a desired installation location in our compile_MET_all.sh script. 

I'm so glad you were able to get a successful compilation when using the script.  Thank you for letting us know.  I'll go ahead and mark an answer for this Discussion, close it and lock it.  Please feel free to start a new Discussion with any other questions you may have when learning to use the tools.
